linera-core/src/client/mod.rs

```diff
@@ -2331,6 +2341,76 @@ impl<Env: Environment> ChainClient<Env> {
         Ok(())
     }
 
+    /// Downloads only the specific sender blocks needed for missing cross-chain messages.
+    /// This is a targeted alternative to `find_received_certificates` that only downloads
+    /// the exact sender blocks we're missing, rather than searching through all received
+    /// certificates.
+    #[instrument(level = "trace")]
+    async fn download_missing_sender_blocks(
+        &self,
+        missing_blocks: BTreeMap<ChainId, Vec<BlockHeight>>,
+    ) -> Result<(), ChainClientError> {
+        if missing_blocks.is_empty() {
+            return Ok(());
+        }
+
+        let (_, committee) = self.admin_committee().await?;
+        let nodes = self.client.make_nodes(&committee)?;
+        let (max_epoch, committees) = self.client.admin_committees().await?;
+
+        // Download certificates for each sender chain at the specific heights.
+        let certificates = stream::iter(missing_blocks.into_iter())
+            .map(|(sender_chain_id, heights)| {
+                let nodes = nodes.clone();
+                async move {
+                    // Try to download from any node.
+                    for node in &nodes {
+                        if let Ok(certs) = node
+                            .download_certificates_by_heights(sender_chain_id, heights.clone())
+                            .await
+                        {
+                            return Ok::<Vec<_>, ChainClientError>(certs);
+                        }
+                    }
+                    // If all nodes fail, return an error.
+                    Err(ChainClientError::CannotDownloadMissingSenderBlocks {
+                        chain_id: sender_chain_id,
+                        heights,
+                    })
+                }
+            })
+            .buffer_unordered(self.options.max_joined_tasks)
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        // Process and validate the downloaded certificates.
+        let mut valid_certificates = Vec::new();
+        for certificate in certificates.into_iter().flatten() {
+            match Client::<Env>::check_certificate(max_epoch, &committees, &certificate)? {
+                CheckCertificateResult::FutureEpoch | CheckCertificateResult::OldEpoch => {
+                    // Skip certificates from unrecognized epochs.
+                    continue;
+                }
+                CheckCertificateResult::New => {
+                    valid_certificates.push(certificate);
+                }
+            }
+        }
+
+        // Receive and process the certificates.
+        for certificate in valid_certificates {
+            self.client
+                .receive_sender_certificate(
+                    certificate,
+                    ReceiveCertificateMode::AlreadyChecked,
+                    Some(nodes.clone()),
+                )
+                .await?;
+        }
```

Comment:
If you extract this part then this function is no longer specific to sender chains and can be reused anywhere where we expect to download lots of certs.

Reply:
Do you have another call site in mind, where this is currently duplicated?

Reply:
In this file there are a couple that try to do the same – `download_certificates`, `download_certificates_from`,